### PR TITLE
Improve FunctionCounter scrape error diagnostics

### PR DESCRIPTION
**Pull request by k-antonov**
## Summary

Improve diagnostics when a Prometheus `FunctionCounter` produces a negative value during scrape. (Issue https://github.com/prometheus/client_java/issues/1090)

Previously, the scrape failed with:

`-2.0: counters cannot have a negative value`

This made it hard to identify which meter was responsible.

This change preserves the existing fail-fast behavior, but enriches the exception message with the Micrometer meter name.

## What changed

- Enrich the `IllegalArgumentException` raised from the `FunctionCounter` Prometheus export path with the meter name
- Add test covering the negative `FunctionCounter` scrape scenario
- Now the exception message looks like `Failed to create Prometheus function counter snapshot for meter 'api.requests': -2.0: counters cannot have a negative value`

## Verification

Verified with a PrometheusMeterRegistryTest and small Spring Boot demo app with this Micrometer branch

### CODE REVIEW

File: implementations/micrometer-registry-prometheus/src/main/java/io/micrometer/prometheusmetrics/PrometheusMeterRegistry.java [hunk 1]
```diff
@@ -380,15 +380,28 @@ protected <T> FunctionCounter newFunctionCounter(Meter.Id id, T obj, ToDoubleFun
...
 
+    private CounterDataPointSnapshot newFunctionCounterDataPointSnapshot(Meter.Id id, double value,
+            List<String> tagKeys, List<String> tagValues, long createdTimestampMillis) {
+        try {
+            return new CounterDataPointSnapshot(value, Labels.of(tagKeys, tagValues), null, createdTimestampMillis);
+        }
+        catch (IllegalArgumentException ex) {
+            if (value < 0.0) {
+                throw new IllegalArgumentException("Failed to create Prometheus function counter snapshot for meter '"
+                        + id.getName() + "': " + ex.getMessage(), ex);
```
Comment (shakuzen) [inline implementations/micrometer-registry-prometheus/src/main/java/io/micrometer/prometheusmetrics/PrometheusMeterRegistry.java]: It's probably more idiomatic to not add the exception message to the message of the exception we will throw since we are passing that exception as the cause. If the stack trace is printed, it will show the message from each, with the cause's message after this exception's message. It ends up duplicating the caught exception's message in the output. What do you think?

**Review by shakuzen — COMMENTED** [review]
Thanks for taking the time to put together a PR. In general, I think it's a nice improvement. However, from the linked issue in the Prometheus client, it seems they are amenable to changing the behavior in the client, which might be at odds or redundant with these changes. Probably that doesn't have to block doing something here but it's something to keep in mind. We should also consider whether to make this change in maintenance branches.

File: implementations/micrometer-registry-prometheus/src/main/java/io/micrometer/prometheusmetrics/PrometheusMeterRegistry.java (hunk 1 above)
Comment (k-antonov) [reply implementations/micrometer-registry-prometheus/src/main/java/io/micrometer/prometheusmetrics/PrometheusMeterRegistry.java]: Thank you, makes sense, updated the PR. I also got rid of strict exception message check, so the test is independent of client message.
I agree we should keep potential client changes in mind. But even if it updates behaviour in this area, in my opinion I suggest a low-risk improvement. In the worst case, the PR may become partially redundant or a candidate for simplification later. However, for now, it will already solve a real user problem.

**Comment by jonatan-ivanov** [comment]
I think this issue is connected: https://github.com/micrometer-metrics/micrometer/issues/7302

File: implementations/micrometer-registry-prometheus/src/main/java/io/micrometer/prometheusmetrics/PrometheusMeterRegistry.java [hunk 2]
```diff
@@ -380,15 +380,28 @@ protected <T> FunctionCounter newFunctionCounter(Meter.Id id, T obj, ToDoubleFun
...
+        }
+        catch (IllegalArgumentException ex) {
+            if (value < 0.0) {
+                throw new IllegalArgumentException(
+                        "Failed to create Prometheus function counter snapshot for meter '" + id.getName() + "'", ex);
+            }
+            throw ex;
+        }
+    }
+
```
Comment (jonatan-ivanov) [inline implementations/micrometer-registry-prometheus/src/main/java/io/micrometer/prometheusmetrics/PrometheusMeterRegistry.java:404]: I have two questions here:
- Should we instead of throwing an exception at scrape time, log a warning and return `NaN` as the value?
- Would not this be cleaner in a separate `PrometheusFunctionCounter` class?

See this as an example for both: https://github.com/jonatan-ivanov/micrometer/commit/8dec1bdde74f2e600604b6615032dc9e7313b373

File: implementations/micrometer-registry-prometheus/src/test/java/io/micrometer/prometheusmetrics/PrometheusMeterRegistryTest.java [hunk 3]
```diff
@@ -450,6 +450,21 @@ void functionCounterNamingConvention() {
...
+        AtomicInteger value = new AtomicInteger(1);
+        FunctionCounter.builder("api.requests", value, AtomicInteger::doubleValue).register(registry);
+
+        assertThat(registry.scrape()).contains("api_requests_total 1.0");
+
+        value.set(-2);
+
+        Throwable thrown = catchThrowable(registry::scrape);
+
+        assertThat(thrown).isInstanceOf(IllegalArgumentException.class).hasMessageContaining("api.requests");
```
Comment (jonatan-ivanov) [inline implementations/micrometer-registry-prometheus/src/test/java/io/micrometer/prometheusmetrics/PrometheusMeterRegistryTest.java:464]: How about something like this? (Maybe with adding a bit more details for the error message?)
```java
assertThatThrownBy(registry::scrape)
    .isInstanceOf(IllegalArgumentException.class)
    .hasMessageContaining("api.requests")
    .hasCauseInstanceOf(IllegalArgumentException.class);
```

File: implementations/micrometer-registry-prometheus/src/main/java/io/micrometer/prometheusmetrics/PrometheusMeterRegistry.java (hunk 2 above)
Comment (jonatan-ivanov) [reply implementations/micrometer-registry-prometheus/src/main/java/io/micrometer/prometheusmetrics/PrometheusMeterRegistry.java:404]: I think this might be a problem for `FunctionTimer` too, not sure if `SummaryDataPointSnapshot` validates `count` and `sum` but I feel it should similarly to `CounterDataPointSnapshot`.

**Review by jonatan-ivanov — COMMENTED** [review]
Thank you for the PR!

File: implementations/micrometer-registry-prometheus/src/main/java/io/micrometer/prometheusmetrics/PrometheusMeterRegistry.java (hunk 2 above)
Comment (k-antonov) [reply implementations/micrometer-registry-prometheus/src/main/java/io/micrometer/prometheusmetrics/PrometheusMeterRegistry.java:404]: Thank you for review!
1) Your https://github.com/jonatan-ivanov/micrometer/commit/8dec1bdde74f2e600604b6615032dc9e7313b373 solution is clean and it resolves the initial issue
2) Unlike `FunctionCounter`, `FunctionTimer` already clamps negative values to 0 in [CumulativeFunctionTimer](https://github.com/micrometer-metrics/micrometer/blob/04096266e526d6ff3817d245ef12477625daca20/micrometer-core/src/main/java/io/micrometer/core/instrument/cumulative/CumulativeFunctionTimer.java#L62-L79) before the `SummaryDataPointSnapshot` is created. I don't see same problem here. Should we add `WarnThenDebugLogger` log to [count and totalTime](https://github.com/micrometer-metrics/micrometer/blob/04096266e526d6ff3817d245ef12477625daca20/micrometer-core/src/main/java/io/micrometer/core/instrument/cumulative/CumulativeFunctionTimer.java#L62-L79)?

File: implementations/micrometer-registry-prometheus/src/test/java/io/micrometer/prometheusmetrics/PrometheusMeterRegistryTest.java (hunk 3 above)
Comment (k-antonov) [reply implementations/micrometer-registry-prometheus/src/test/java/io/micrometer/prometheusmetrics/PrometheusMeterRegistryTest.java:464]: I agree, I would accept this if we still needed this test

Comment (trip-pguedes) [reply implementations/micrometer-registry-prometheus/src/test/java/io/micrometer/prometheusmetrics/PrometheusMeterRegistryTest.java:464]: Bumping this as we still have the problem reported in https://github.com/micrometer-metrics/micrometer/issues/7302
